src/python/pants/backend/python/tasks/resolve_requirements_task_base.py

```diff
@@ -62,13 +64,43 @@ def resolve_requirements(self, req_libs, local_dist_targets=None):
           # Handle locally-built python distribution dependencies.
           built_dists = self.context.products.get_data(BuildLocalPythonDistributions.PYTHON_DISTS)
           if built_dists:
-            req_libs = inject_synthetic_dist_requirements(self.context.build_graph,
-                                                          built_dists,
-                                                          ':'.join(2 * [target_set_id])) + req_libs
-          self._build_requirements_pex(interpreter, safe_path, req_libs)
+            req_libs = inject_synthetic_dist_requirements(
+              self.context.build_graph, built_dists, ':'.join(2 * [target_set_id])
+            ) + req_libs
+          builder = PEXBuilder(path=safe_path, interpreter=interpreter, copy=True)
+          dump_requirement_libs(builder, interpreter, req_libs, self.context.log)
+          builder.freeze()
     return PEX(path, interpreter=interpreter)
 
-  def _build_requirements_pex(self, interpreter, path, req_libs):
-    builder = PEXBuilder(path=path, interpreter=interpreter, copy=True)
-    dump_requirements(builder, interpreter, req_libs, self.context.log)
-    builder.freeze()
+  def resolve_requirement_strings(self, interpreter, requirement_strings):
+    """Resolve a list of pip-style requirement strings."""
+    requirement_strings = sorted(requirement_strings)
+    if len(requirement_strings) == 0:
+      req_strings_id = 'no_requirements'
+    elif len(requirement_strings) == 1:
+      req_strings_id = requirement_strings[0]
+    else:
+      req_strings_id = hash_all(requirement_strings)
+
+    path = os.path.realpath(os.path.join(self.workdir, str(interpreter.identity), req_strings_id))
```

Comment:
It appears that this "extra_requirements" pex was previously governed by the `CacheManager` + an invalidation check. I am not sure if hashing in this way would cause any issues since the reqs are coming from `self.extra_requirements` (which don't/rarely change afaict?). Is the invalidation strategy a concern here?

Reply:
The id will change if any of the requirements strings changes, and the requirements will be re-resolved.  So this is as robust as generating a synthetic target and running invalidation on it, without the superfluous machinery. We have to create a unique output path anyway, so we couldn't remove any of this code even if we did do the other thing.